vortex-serde/src/io/tokio.rs

```diff
@@ -38,30 +43,100 @@ impl<W: AsyncWrite + Unpin> VortexWrite for TokioAdapter<W> {
-        let std_file = self.try_clone().await?.into_std().await;
-        std_file.read_exact_at(buffer.as_mut(), pos)?;
-        Ok(buffer)
+impl AsyncRuntime for Handle {
+    fn block_on<F: Future>(&self, fut: F) -> F::Output {
+        self.block_on(fut)
+    }
+}
+
+/// A cheaply cloneable, readonly file that executes operations
+/// on a tokio blocking threadpool.
+///
+/// We use this because the builtin tokio `File` type is not `Clone` and
+/// also does actually implement a `read_exact_at` operation.
+#[derive(Debug, Clone)]
+pub struct TokioFile(Arc<Inner>);
+
+// Implement an inner file.
+#[derive(Debug)]
+struct Inner {
+    file: ManuallyDrop<File>,
+}
+
+impl TokioFile {
+    /// Open a file on the current file system.
+    ///
+    /// The `TokioFile` takes ownership of the file descriptor, and can be cloned
+    /// many times without opening a new file descriptor. When the last instance
+    /// of the `TokioFile` is dropped, the file descriptor is closed.
+    pub fn open(path: impl AsRef<Path>) -> io::Result<Self> {
+        let f = File::open(path)?;
+        let inner = Arc::new(Inner {
+            file: ManuallyDrop::new(f),
+        });
+
+        Ok(Self(inner))
+    }
+}
+
+// Implement deref coercion for non-mut `File` methods on `TokioFile`.
+impl Deref for TokioFile {
+    type Target = File;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0.file
     }
+}
+
+impl VortexReadAt for TokioFile {
+    fn read_at_into(
+        &self,
+        pos: u64,
+        buffer: BytesMut,
+    ) -> impl Future<Output = io::Result<BytesMut>> + 'static {
+        let this = self.clone();
 
-    async fn size(&self) -> u64 {
-        self.metadata()
-            .await
-            .map_err(|err| VortexError::IOError(err).with_context("Failed to get file metadata"))
-            .vortex_unwrap()
-            .len()
+        async move {
+            let res = tokio::task::spawn_blocking(move || {
+                let mut buffer = buffer;
+                match this.read_exact_at(&mut buffer, pos) {
+                    Ok(()) => Ok(buffer),
+                    Err(e) => Err(e),
+                }
+            });
+
+            res.await.map_err(|_canceled| {
+                io::Error::new(
+                    io::ErrorKind::Other,
+                    "blocking read_exact_at task was canceled",
+                )
+            })?
+        }
+    }
+
+    fn size(&self) -> impl Future<Output = u64> + 'static {
+        let this = self.clone();
+
+        async move {
+            let res = tokio::task::spawn_blocking(move || {
+                this.metadata()
+                    .unwrap_or_else(|e| vortex_panic!("access TokioFile metadata: {e}"))
+                    .len()
+            })
+            .await;
+
+            res.unwrap_or_else(|e| vortex_panic!("Joining spawn_blocking: size: {e}"))
+        }
     }
 }
 
-impl VortexWrite for File {
+impl VortexWrite for tokio::fs::File {
```

Comment:
`VortexWrite` is stateful and takes `mut` access, so it cannot be shared so we can use tokio's builtin `File` type instead of `TokioFile`. An argument could be made that we should only ever use `TokioFile` for consistency but idk